### PR TITLE
Status history

### PR DESCRIPTION
**Pull request by SimonRichardson**
The status history is now a dependency for the service factory. This
allows us to set history at the service level. The status history
worker uses the logsink worker as the backing store. By using an
intermediary we can change the backend to the status history without
updating the dependencies of each.

The roadblock to this change was that the model logger required
the model uuid, model owner and model name. There was a latent bug
where if you changed the model owner and model name, but used the
same model uuid it would give you a cached model logger to file you
might not want.

The logsink also can't depend on the service factory because we want
to use the logsink in the service factory. To break the cycle, I
pushed the controller config values into the agent config. This is
a common pattern that we've used before. The prior logsink worker
didn't update the logger size and rotation information if it changed,
only if the agent was rebooted. I've mirrored what we currently
have, although it's relatively simple to update the agent config
updater to include those values, which would do it automatically.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd test
$ juju add-model default
$ juju deploy ubuntu
```

We should see the statushistory log into the model log.

```sh
$ juju ssh -m controller 0
> cat /var/log/juju/models/<model path>.log
```

## Links

**Jira card:** [JUJU-5888](https://warthogs.atlassian.net/browse/JUJU-5888)

[JUJU-5888]: https://warthogs.atlassian.net/browse/JUJU-5888?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by hpidcock — COMMENTED** [review]
Looks great. Only a few minor issues.

File: domain/model/service/service.go [hunk 1]
```diff
@@ -206,6 +209,9 @@ func (s *Service) CreateModel(
 	}
 
 	finaliser := ModelFinaliser(func(ctx context.Context) error {
+		setter := s.statusHistoryFactory.StatusHistorySetterForModel(uuid.String())
+		setter.SetStatusHistory(status.KindModel, status.Available, uuid.String())
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply domain/model/service/service.go:213]: I'll fix.

File: internal/worker/logsink/loggers.go [hunk 2]
```diff
@@ -34,14 +35,21 @@ func NewModelLogger(
...
+	modelLogger := &modelLogger{
 		clock:               clock,
 		loggerBufferSize:    bufferSize,
 		loggerFlushInterval: flushInterval,
 		loggerForModel:      loggerForModelFunc,
 		modelLoggers:        make(map[string]corelogger.LoggerCloser),
 	}
+
+	// Create the fallback logger for models that have not been initialized yet.
+	if err := modelLogger.initLogger(database.ControllerNS, "log", "admin"); err != nil {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply internal/worker/logsink/loggers.go:48]: Legit.

File: internal/worker/statushistory/manifold_test.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
+// Copyright 2023 Canonical Ltd.
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply internal/worker/statushistory/manifold_test.go:1]: I've traveled back in time.

File: internal/worker/logsink/loggers_test.go [hunk 4]
```diff
@@ -37,26 +38,59 @@ func (s *stubLogger) Close() error {
...
 	}
-	ml := NewModelLogger(
+	ml, err := NewModelLogger(
 		func(modelUUID, modelName string) (corelogger.LoggerCloser, error) {
 			if l, ok := loggers[modelUUID]; ok {
 				return l, nil
 			}
 			return nil, errors.NotFound
 		},
-		1, time.Millisecond, testclock.NewDilatedWallClock(time.Millisecond),
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply internal/worker/logsink/loggers_test.go:54]: Yeah, I was unsure what it was testing at first.

**Comment by hpidcock** [comment]
```
20:29:17 # github.com/juju/juju/domain/secret_test [github.com/juju/juju/domain/secret.test]
20:29:17 domain/secret/watcher_test.go:36:51: not enough arguments in call to applicationservice.NewService
20:29:17 	have (*"github.com/juju/juju/domain/application/state".State, "github.com/juju/juju/testing".CheckLogger, nil)
20:29:17 	want ("github.com/juju/juju/domain/application/service".State, "github.com/juju/juju/internal/storage".ProviderRegistry, status.StatusHistoryForModel, "github.com/juju/juju/domain/application/service".Logger)
```

**Comment by hpidcock** [comment]
For some reason I'm only seeing the initial status history for only the application
```
{"model-uuid":"e3cfb5dd-16f0-408c-8fbd-7a72b98299a4","timestamp":"2024-04-23T12:20:16.760673033Z","entity":"u2","level":"INFO","module":"statushistory","location":"worker.go:102","message":"status history: application - unset status","labels":{"domain":"status","id":"u2","kind":"application","status":"unset"}}
```
Even if I do something like `juju exec --unit u2/0 -- status-set --application waiting wow4` it doesn't show up in the model's log.

I also see a single log entry in a file named `admin-log-contro.log` with
```
{"model-uuid":"5f165e1c-9ab9-4081-8605-17eb9b33ed7c","timestamp":"2024-04-23T12:06:56.446121818Z","entity":"controller","level":"INFO","module":"statushistory","location":"worker.go:102","message":"status history: application - unset status","labels":{"domain":"status","id":"controller","kind":"application","status":"unset"}}
```

**Review by jameinel — COMMENTED** [review]
I haven't gotten through the whole thing, it is a bit bigger than I was hoping for, we probably should talk about it.

**Comment by SimonRichardson** [comment]
Closing this, for now, will re-open if we want to go that direction.

**Pull request closed without merging**